src/main/java/org/spongepowered/common/registry/type/item/ItemStackComparatorRegistryModule.java

```diff
@@ -0,0 +1,132 @@
+import org.spongepowered.api.registry.RegistryModule;
+import org.spongepowered.api.registry.util.RegisterCatalog;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+
+@SuppressWarnings("unchecked")
+public final class ItemStackComparatorRegistryModule implements RegistryModule {
+
+    @RegisterCatalog(ItemStackComparators.class)
+    private final Map<String, Comparator<ItemStack>> comparators = new HashMap<>();
+
+    @Override
+    public void registerDefaults() {
+
+        Comparator<ItemStack> TYPE = Comparator.comparing(i -> i.getType().getId());
+        this.comparators.put("TYPE", TYPE);
+        Comparator<ItemStack> SIZE = Comparator.comparing(ItemStack::getQuantity);
+        this.comparators.put("SIZE", SIZE);
+        Comparator<ItemStack> TYPE_SIZE = TYPE.thenComparing(SIZE);
+        this.comparators.put("TYPE_SIZE", TYPE_SIZE);
+        this.comparators.put("DEFAULT", TYPE_SIZE);
+        Properties PROPERTIES = new Properties();
+        this.comparators.put("PROPERTIES", PROPERTIES);
+        ItemDataComparator ITEM_DATA = new ItemDataComparator();
+        this.comparators.put("ITEM_DATA", ITEM_DATA);
+        this.comparators.put("ITEM_DATA_IGNORE_DAMAGE", new ItemDataComparator(DamageableData.class));
+        this.comparators.put("IGNORE_SIZE", TYPE.thenComparing(PROPERTIES).thenComparing(ITEM_DATA));
+        this.comparators.put("ALL", TYPE.thenComparing(SIZE).thenComparing(PROPERTIES).thenComparing(ITEM_DATA));
+    }
+
+    static final class Properties implements Comparator<ItemStack> {
+
+        @Override
+        public int compare(@Nullable final ItemStack o1, @Nullable final ItemStack o2) {
+            if (o1 == null && o2 == null) {
+                return 0;
+            }
+            if (o1 == null) {
+                return 1;
+            }
+            if (o2 == null) {
+                return -1;
+            }
+            List<Property<?, ?>> properties = Lists.newArrayList(o2.getApplicableProperties());
+            for (Property<?, ?> property : o1.getApplicableProperties()) {
+                if (properties.contains(property)) {
+                    properties.remove(property);
+                } else {
+                    return -1;
+                }
+            }
+            return properties.size();
+        }
+    }
+    static final class ItemDataComparator implements Comparator<ItemStack> {
+
+        private final Class<? extends DataManipulator<?, ?>>[] ignored;
+
+        public ItemDataComparator(Class<? extends DataManipulator<?, ?>>... ignored) {
+            this.ignored = ignored;
+        }
+
+        @Override
+        public int compare(@Nullable final ItemStack o1, @Nullable final ItemStack o2) {
+            if (o1 == null && o2 == null) {
+                return 0;
+            }
+            if (o1 == null) {
+                return 1;
+            }
+            if (o2 == null) {
```

Comment:
Why are nulls the biggest?

Reply:
copy paste
But i'll remove the null checks.
Itemstacks are not supposed to be null they are ItemStack.empty() now.